src/datasets/arrow_dataset.py

```diff
@@ -425,17 +425,26 @@ def __setstate__(self, state):
         if self._indices is None and self._indices_data_files:
             self._indices = reader._read_files(self._indices_data_files)
 
-    def save_to_disk(self, dataset_path: str):
+    def save_to_disk(
+        self, dataset_path: str, aws_profile="default", aws_access_key_id=None, aws_secret_access_key=None
+    ):
         """
-        Save the dataset in a dataset directory
+        Save the dataset in a dataset directory or to a s3 bucket
 
         Args:
-            dataset_path (``str``): path of the dataset directory where the dataset will be saved to
+            dataset_path (``str``): path or s3 uri of the dataset directory where the dataset will be saved to
+            aws_profile (:obj:`str`,  `optional`, defaults to :obj:``default``): the aws profile used to create the `boto_session` for uploading the data to s3
+            aws_access_key_id (:obj:`str`,  `optional`, defaults to :obj:``None``): the aws access key id used to create the `boto_session` for uploading the data to s3
+            aws_secret_access_key (:obj:`str`,  `optional`, defaults to :obj:``None``): the aws secret access key used to create the `boto_session` for uploading the data to s3
```

Comment:
if we have `boto3` as an optional dependency, maybe there's a way to not use those (verbose) params and use like a profile name or something instead?

Reply:
As @lhoestq mentioned, we can probably avoid using these params by letting the user provide a custom `fs` directly. I think this has several advantages (avoid having too many params, let's remove code specific to s3, ..)

Reply:
so you would suggest something like that? 
```python
import fsspec

s3 = fsspec.filesystem("s3", anon=False, key=aws_access_key_id, secret=aws_secret_access_key)

dataset.save_to_disk('s3://my-s3-bucket-with-region/dataset/train',fs=s3)
```

What I don't like about that is the customer experience. `fsspec` is not that well documented for the remote filesystems, e.g. when you want to know which "credentials" you need you to have to go to the `s3fs` documentation.

what do you think if we remove the named arguments `aws_profile`... and handle it how `fsspec` does with an `storage_options` dict. 
```python
dataset.save_to_disk('s3://my-s3-bucket-with-region/dataset/train',
                                      storage_options={
                                             'aws_access_key_id': 123,
                                             'aws_secret_access_key': 123
                                      })
```

Reply:
Indeed the docstring of `fsspec.filesystem` is not ideal:
```
Signature: fsspec.filesystem(protocol, **storage_options)
Docstring:
Instantiate filesystems for given protocol and arguments

``storage_options`` are specific to the protocol being chosen, and are
passed directly to the class.
```
Maybe we can have a better documentation on our side instead using a wrapper:
```python
import datasets

fs = datasets.filesystem("s3", anon=False, key=aws_access_key_id, secret=aws_secret_access_key)
```
Where the docstring of `datasets.filesystem` is more complete and includes examples for popular filesystems like s3

Reply:
Another option would be to make available the filesystems classes easily:
```python
>>> from datasets.filesystems import S3FileSystem  # S3FileSystem si simply the class from s3fs
>>> S3FileSystem?  # show the docstring
```
shows
```
Init signature: s3fs.core.S3FileSystem(*args, **kwargs)
Docstring:     
Access S3 as if it were a file system.

This exposes a filesystem-like API (ls, cp, open, etc.) on top of S3
storage.

Provide credentials either explicitly (``key=``, ``secret=``) or depend
on boto's credential methods. See botocore documentation for more
information. If no credentials are available, use ``anon=True``.

Parameters
----------
anon : bool (False)
    Whether to use anonymous connection (public buckets only). If False,
    uses the key/secret given, or boto's credential resolver (client_kwargs,
    environment, variables, config files, EC2 IAM server, in that order)
key : string (None)
    If not anonymous, use this access key ID, if specified
secret : string (None)
    If not anonymous, use this secret access key, if specified
token : string (None)
    If not anonymous, use this security token, if specified
etc.
```

Reply:
- If we add the various params separately, the more new filesystems we support, the more complicated it becomes. For the user it becomes difficult to know which params to look at, and for us to document everything. It seems like a good option to test things out, but having to change it down the road will require breaking changes, and we usually try to avoid these as much as possible.
- Using some kind of `storage_options` just like `fsspec.filesystem` might be better but it seems difficult to document also. I think the same argument applies to having a `datasets.filesystem` helper.

I think I have a preference for your second option @lhoestq 

- It seems easy to show all the available filesystems to the user, with each of them having a meaningful documentation
- We can probably add tests for those we support explicitly
- If all of them share the same interface, then power users can probably use anything they want from `fsspec` without explicit support from us?

What do you guys think?

Reply:
Yes I think the second option is interesting and I totally agree with your three points.
Maybe let's start by having `S3FileSystem ` in `datasets.filesystems` and we can add the other ones later.

In the documentation of `save_to_disk`/`load_from_disk` we can then say that any filesystem from `datasets.filesystems` or `fsspec` can be used.

Reply:
I have rebuilt everything so that you can now pass in a `fsspec` like filesystem.

```python 
from datasets import S3Filesystem, load_from_disk

s3 = datasets.S3FileSystem(key=aws_access_key_id, secret=aws_secret_access_key)  # doctest: +SKIP

dataset = load_from_disk('s3://my-private-datasets/imdb/train',fs=s3)  # doctest: +SKIP

print(len(dataset))
# 25000
```

I also created a "draft" documentation version with a few examples for `S3Fileystem`. It would be nice if take look at everything. Afterwards, I would adjust the documentation of `save_to_dist`/`load_from_disk` 

![image](https://user-images.githubusercontent.com/32632186/104847565-12ba0980-58e1-11eb-8b2a-ee30011c304c.png)